design/src/rfcs/rfc0024_better_constraint_violations.md

```diff
@@ -0,0 +1,509 @@
+    fn try_from(value: std::string::String) -> Result<Self, Self::Error> {
+        Self::check_length(&value)?;
+
+        let value = Self::check_pattern(value)?;
+
+        Ok(Self(value))
+    }
+}
+```
+
+Observe how a failure to adhere to the `@length` trait will short-circuit the
+evaluation of the constructor, when the value could technically also not adhere
+with the `@pattern` trait.
+
+Similarly, constrained structures fail upon encountering the first member that
+violates a constraint.
+
+Additionally, _in framework request deserialization code_:
+
+- collections whose members are constrained fail upon encountering the first
+  member that violates the constraint; and
+- maps whose keys and/or values are constrained fail upon encountering the
+  first violation; and
+
+In summary, any shape that is transitively constrained yields types whose
+constructors (both the internal one and the user-facing one) currently
+short-circuit upon encountering the first violation.
+
+### Solution proposal
+
+The deserializing architecture lends itself to be easily refactored so that we
+can collect constraint violations before returning them. Indeed, note that
+deserializers enforce constraint traits in a two-step phase: first, the
+_entirety_ of the unconstrained value is deserialized, _then_ constraint traits
+are enforced by feeding the entire value to the `TryFrom` constructor.
+
+We will introduce a `ConstraintViolations` type (note the plural) that
+represents a collection of constraint violations that can occur _within user
+application code_. Roughly:
+
+```rust
+pub ConstraintViolations<T>(pub(crate) Vec<T>);
+
+impl<T> IntoIterator<Item = T> for ConstraintViolations<T> { ... }
+
+impl std::convert::TryFrom<std::string::String> for LengthPatternString {
+    type Error = ConstraintViolations<crate::model::length_patterh_string::ConstraintViolation>;
+
+    fn try_from(value: std::string::String) -> Result<Self, Self::Error> {
+        // Check constraints and collect violations.
+        ...
+    }
+}
+```
+
+- The main reason for wrapping a vector in `ConstraintViolations` as opposed to
+  directly returning the vector is forwards-compatibility: we may want to
+  expand `ConstraintViolations` with conveniences.
+- If the constrained type can only ever yield a single violation, we will
+  dispense with `ConstraintViolations` and keep directly returning the
+  `crate::model::shape_name::ConstraintViolation` type.
+
+We will analogously introduce a `ConstraintViolationExceptions` type that
+represents a collection of constraint violations that can occur _within the
+framework's request deserialization code_. This type will be `pub(crate)` and
+will be the one the framework will map to Smithy's `ValidationException` that
+eventually gets serialized into the response.
+
+### Unresolved questions
+
+#### Collecting constraint violations may constitute a DOS attack vector
```

Comment:
Some comments to pushing back against https://github.com/awslabs/smithy-rs/pull/1199#discussion_r809300673.

Shouldn't clients be in the business of validating the requests they provide?

- Web forms can do it using a client side script - this is better UX because they get an immediate response.
- Typed languages can do it via fallible request constructors - again, this is a better UX.

Currently, there are three logical phases:

1. Receive bytes into a buffer.
2. Deserialize the buffer into the unconstrained tree.
3. Convert the unconstrained tree into the constrained tree.

Phase (2) requires lots of allocations - copying bytes into `String` for example. Phase (3) also requires allocations, `into_iter()...collect()` dynamic collections and some extra CPU work in the case of `@pattern` etc.

If we don't return early on an error we incur all the allocations related to (2) for invalid requests. Validation doesn't protect the parser if we parse then validate, but if we "parse don't validate" (deserialize directly into the constrained tree) it does to some extent.

If we deserialize the buffer directly into the constrained tree we are also able to avoid the allocations associated with phase 3.

Reply:
> Shouldn't clients be in the business of validating the requests they provide?

Yes, but client-side validation is best effort and a backend service should not rely on it for data integrity.

---

I'm also unconvinced of the value of collecting constraint violations in the general use cases of smithy-rs server SDKs. But there's certainly particular use cases that need it. Another one that I failed to mention in the document is backwards compatibility. If a service has always returned a custom response with a _list_ of validation error messages, this feature and [custom validation error messages](https://smithy.io/ts-ssdk/validation.html#custom-validation) will enable them to move their service to use a smithy-rs sSDK.

Regarding deserializing directly into the constrained tree, that's a bigger architectural change that was already expressed. Deserializing directly into the constrained tree while collecting constraint violations at the same time is difficult. [My preference](https://github.com/awslabs/smithy-rs/pull/1342#discussion_r1012777985) is to explore it in a completely separate deserialization approach by implementing `serde` visitors.

Reply:
> Yes, but client-side validation is best effort and a backend service should not rely on it for data integrity.

For sure, we shouldn't forgo server side validation and hope that clients behave. I'm proposing that server side validation returns early and client side validation can collect all errors if it improves UX for them.

I'd argue that this is the norm in most client-server interactions - if a server has a list of criteria that a request needs to meet then the client should expect an error message describing the first criteria which fails, rather than a list of all violations.

> Regarding deserializing directly into the constrained tree, that's a bigger architectural change that was already expressed. Deserializing directly into the constrained tree while collecting constraint violations at the same time is difficult. https://github.com/awslabs/smithy-rs/pull/1342#discussion_r1012777985 is to explore it in a completely separate deserialization approach by implementing serde visitors.

Agree, I'm mentioning it because I think "deserializing directly into the constrained tree" is easiest if we return errors early while "Deserializing directly into the constrained tree while collecting constraint violations" might be impossible within `serde`.